dbms/src/Storages/MergeTree/IMergedBlockOutputStream.cpp

```diff
@@ -33,4 +38,78 @@ Block IMergedBlockOutputStream::getBlockAndPermute(const Block & block, const Na
     return result;
 }
 
+NameSet IMergedBlockOutputStream::removeEmptyColumnsFromPart(
+    const MergeTreeDataPartPtr & data_part,
+    NamesAndTypesList & columns,
+    MergeTreeData::DataPart::Checksums & checksums)
+{
+    const NameSet & empty_columns = data_part->expired_columns;
+
+    /// For compact part we have to override whole file with data, it's not
+    /// worth it
+    if (empty_columns.empty() || isCompactPart(data_part))
+        return {};
+
+    /// Collect counts for shared streams of different columns. As an example, Nested columns have shared stream with array sizes.
+    std::map<String, size_t> stream_counts;
+    for (const NameAndTypePair & column : columns)
+    {
+        column.type->enumerateStreams(
+            [&](const IDataType::SubstreamPath & substream_path)
+            {
+                ++stream_counts[IDataType::getFileNameForStream(column.name, substream_path)];
+            },
+            {});
+    }
+
+    NameSet remove_files;
+    const String mrk_extension = data_part->getMarksFileExtension();
+    for (const auto & column_name : empty_columns)
+    {
+        IDataType::StreamCallback callback = [&](const IDataType::SubstreamPath & substream_path)
+        {
+            String stream_name = IDataType::getFileNameForStream(column_name, substream_path);
+            /// Delete files if they are no longer shared with another column.
+            if (--stream_counts[stream_name] == 0)
+            {
+                remove_files.emplace(stream_name + ".bin");
+                remove_files.emplace(stream_name + mrk_extension);
+            }
+        };
+        IDataType::SubstreamPath stream_path;
+        auto column_with_type = columns.tryGetByName(column_name);
+        if (column_with_type)
+            column_with_type->type->enumerateStreams(callback, stream_path);
+    }
+
+    /// Remove files on disk and checksums
+    for (const String & removed_file : remove_files)
+    {
+        if (checksums.files.count(removed_file))
+        {
+            String path_to_remove = data_part->getFullPath() + removed_file;
+
+            if (0 != unlink(path_to_remove.c_str()))
```

Comment:
Should go through disk interface.
disk->remove(data_part->getFullRelativePath() + removed_file)